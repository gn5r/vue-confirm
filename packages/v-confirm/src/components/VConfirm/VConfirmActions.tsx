// Vue
import { defineComponent, toRef, computed } from "vue";

// Composables
import { useBackgroundColor, useTextColor } from "@/composables/color";
import { useBtnProps } from "@/composables/btn";

// Types
import type { ButtonHTMLAttributes } from "vue";

export const VConfirmActions = defineComponent({
  name: "v-confirm-footer",
  props: {
    ...useBtnProps(),
  },
  setup(props, { slots }) {
    const items = computed(() =>
      props.btns.map((btn) => {
        const { backgroundColorClasses, backgroundColorStyles } =
          useBackgroundColor(toRef(btn, "color"));
        const { textColorClasses, textColorStyles } = useTextColor(
          toRef(btn, "textColor")
        );
        const classes: string[] = [];
        if (btn.class) {
          if (Array.isArray(btn.class)) classes.push(...btn.class);
          else classes.push(btn.class);
        } else {
          classes.push("button");
        }
        return {
          attrs: {
            type: "button" as ButtonHTMLAttributes["type"],
            class: [
              ...classes,
              ...backgroundColorClasses.value,
              ...textColorClasses.value,
            ],
            style: { ...backgroundColorStyles.value, ...textColorStyles.value },
            onClick: btn.onClick,
          },
          ...btn,
        };
      })
    );

    return () => (
      <div
        class="v-confirm__actions"
        style={{ justifyContent: props.btnAlign }}
      >
        {slots.default
          ? slots.default({ items: items.value })
          : items.value.map((item) => {
              return <button {...item.attrs}>{item.text}</button>;
            })}
      </div>
    );
  },
});

export type VConfirmActions = InstanceType<typeof VConfirmActions>;
