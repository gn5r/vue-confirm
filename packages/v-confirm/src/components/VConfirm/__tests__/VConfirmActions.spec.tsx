import { mount } from "@vue/test-utils";
import { describe, it, expect, vi } from "vitest";
import { VConfirmActions } from "../VConfirmActions";
import type { VConfirmBtn } from "@gn5r/vue-confirm";

describe("VConfirmActions.tsx", () => {
  it("render a component", () => {
    const wrapper = mount(VConfirmActions);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("not render a button when btns is empty", () => {
    const wrapper = mount(VConfirmActions);
    expect(wrapper.find(".v-confirm__actions > button").exists()).toBe(false);
  });

  it("should have justify-content value of 'end' in div's style when using default props", () => {
    const wrapper = mount(VConfirmActions);
    expect(wrapper.find(".v-confirm__actions").attributes().style).toContain(
      "justify-content: end"
    );
  });

  it("should set justify-content value to btnAlign value when specified in props", () => {
    const wrapper = mount(VConfirmActions, { props: { btnAlign: "start" } });
    expect(wrapper.find(".v-confirm__actions").attributes().style).toContain(
      "justify-content: start"
    );
  });

  it("render buttons", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        onClick: vi.fn(),
      },
      {
        text: "NO",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button").exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("render a text in button", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button").text()).toBe("OK");
  });

  it("should default class is .button of button", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button.button").exists()).toBe(
      true
    );
  });

  it("should set button class to the specified class value in btns", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        class: "btn",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button.button").exists()).toBe(
      false
    );
    expect(wrapper.find(".v-confirm__actions > button.btn").exists()).toBe(
      true
    );
  });

  it("render a button with class style background color", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        color: "primary",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button").classes()).toContain(
      "primary"
    );
  });

  it("render a button with css style background color", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        color: "#fff",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(
      wrapper.find(".v-confirm__actions > button").attributes().style
    ).toContain("background-color: rgb(255, 255, 255)");
  });

  it("render a button with class style color", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        textColor: "white--text",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(wrapper.find(".v-confirm__actions > button").classes()).toContain(
      "white--text"
    );
  });

  it("render a button with css style color", () => {
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        textColor: "#fff",
        onClick: vi.fn(),
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    expect(
      wrapper.find(".v-confirm__actions > button").attributes().style
    ).toContain("color: rgb(255, 255, 255); caret-color: #fff");
  });

  it("should execute button function when button is clicked", async () => {
    const btnFunc = vi.fn();
    const btns: VConfirmBtn[] = [
      {
        text: "OK",
        onClick: btnFunc,
      },
    ];
    const wrapper = mount(VConfirmActions, { props: { btns: btns } });
    await wrapper.find(".v-confirm__actions > button").trigger("click");
    expect(btnFunc).toHaveBeenCalledTimes(1);
  });
});
