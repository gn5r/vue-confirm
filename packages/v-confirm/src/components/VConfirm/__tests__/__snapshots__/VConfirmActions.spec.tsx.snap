// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VConfirmActions.tsx > render a component 1`] = `"<div class=\\"v-confirm__actions\\" style=\\"justify-content: end;\\"></div>"`;

exports[`VConfirmActions.tsx > render buttons 1`] = `"<div class=\\"v-confirm__actions\\" style=\\"justify-content: end;\\"><button type=\\"button\\" class=\\"button\\">OK</button><button type=\\"button\\" class=\\"button\\">NO</button></div>"`;
